src/event.rs

```diff
@@ -239,94 +221,157 @@ pub enum WindowEvent<'a> {
 
-    /// The cursor has entered the window.
-    CursorEntered { device_id: DeviceId },
+    PointerEntered(PointerId),
+    PointerLeft(PointerId),
 
-    /// The cursor has left the window.
-    CursorLeft { device_id: DeviceId },
+    PointerForce(PointerId, Force),
 
-    /// A mouse wheel movement or touchpad scroll occurred.
-    MouseWheel {
-        device_id: DeviceId,
-        delta: MouseScrollDelta,
-        phase: TouchPhase,
-        #[deprecated = "Deprecated in favor of WindowEvent::ModifiersChanged"]
-        modifiers: ModifiersState,
-    },
+    PointerPress(PointerId, PointerPress),
 
-    /// An mouse button press has been received.
-    MouseInput {
-        device_id: DeviceId,
-        state: ElementState,
-        button: MouseButton,
-        #[deprecated = "Deprecated in favor of WindowEvent::ModifiersChanged"]
-        modifiers: ModifiersState,
-    },
+    ScrollStarted,
+    ScrollDiscrete(Vector<i32>),
+    ScrollSmooth(Vector<f64>),
+    ScrollEnded,
 
-    /// Touchpad pressure event.
+    /// The system window theme has changed.
     ///
-    /// At the moment, only supported on Apple forcetouch-capable macbooks.
-    /// The parameters are: pressure level (value between 0 and 1 representing how hard the touchpad
-    /// is being pressed) and stage (integer representing the click level).
-    TouchpadPressure {
-        device_id: DeviceId,
-        pressure: f32,
-        stage: i64,
-    },
+    /// Applications might wish to react to this to change the theme of the content of the window
+    /// when the system changes the window theme.
+    ///
+    /// At the moment this is only supported on Windows.
+    ThemeChanged(Theme),
+}
 
-    /// Motion on some analog axis. May report data redundant to other, more specific events.
-    AxisMotion {
-        device_id: DeviceId,
-        axis: AxisId,
-        value: f64,
-    },
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Vector<T> {
+    pub x: T,
+    pub y: T,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct KeyPress {
+    pub(crate) logical_key: Option<LogicalKey>,
+    pub(crate) scan_code: u32,
+    pub(crate) is_down: bool,
+    pub(crate) repeat_count: u32,
+    pub(crate) is_synthetic: bool,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RawKeyPress {
+    pub(crate) logical_key: Option<LogicalKey>,
+    pub(crate) scan_code: u32,
+    pub(crate) is_down: bool,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct PointerPress {
+    pub(crate) button: u8,
+    pub(crate) is_down: bool,
+    pub(crate) click_count: u32,
+}
 
-    /// Touch event has been received
-    Touch(Touch),
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RawPointerPress {
+    pub(crate) button: u8,
+    pub(crate) is_down: bool,
```

Comment:
What's the benefit of disabling direct access to those fields for crate users. I feel like it should be fine and much simpler than calling getters.

Reply:
Direct field access is definitely very convenient. Getters have a couple significant advantages:
- Adding new getters isn't a breaking change, so new data to events is very unlikely to require existing users to update their code.
- Getters can return values that are computed from, rather than stored in, the struct. This is useful when the type of a field need to change to carry additional information, in which case the old getter can be preserved at zero cost, giving users time to switch. Although if fields can be marked deprecated, this might not be a major advantage, since one more field in an event isn't likely to be a significant cost.

Reply:
>Adding new getters isn't a breaking change, so new data to events is very unlikely to require existing users to update their code

I feel like it's related to a fact, that you'll make a constructing from fields available? The thing is that using getters in match arms is just a pain and make the code more nested and complex, however with direct access it's way easier to write things and it makes more sense in the end.

>Getters can return values that are computed from, rather than stored in, the struct. This is useful when the type of a field need to change to carry additional information, in which case the old getter can be preserved at zero cost, giving users time to switch. Although if fields can be marked deprecated, this might not be a major advantage, since one more field in an event isn't likely to be a significant cost.

In general, but we're talking about events here, which you just forward to the user anyway.

Like I just don't feel like it'll make any advantage at all for the end user. You can see that this feature is very heavily used here https://github.com/alacritty/alacritty/blob/058b036fe7f3a7cdc2ebdc21ab25aded783e4e59/alacritty/src/event.rs#L573 , with getters it'll be way harder to write.

Reply:
> I feel like it's related to a fact, that you'll make a constructing from fields available?

I'm not sure what this means. For the same compatibility reasons, these events don't and shouldn't have public constructors.

> The thing is that using getters in match arms is just a pain and make the code more nested and complex, however with direct access it's way easier to write things and it makes more sense in the end.

A side-by-side example was presented above:
```
WindowEvent::KeyPress { key: Some(LogicalKey::A), is_down: true, .. } => ()
WindowEvent::KeyPress(e) if e.key() == Some(LogicalKey::A) && e.is_down() => ()
```
The forwards-compatible version is exactly four characters longer, and I don't think it's in any sense "more nested". Further, in my experience you're only matching on exact values in toy code, so the impact of this is limited.

> In general, but we're talking about events here, which you just forward to the user anyway.

Whether you're matching on a field or forwarding it elsewhere, having its name/type/presence not change when new capabilities are added reduces breakage.

Reply:
> You can see that this feature is very heavily used here https://github.com/alacritty/alacritty/blob/058b036fe7f3a7cdc2ebdc21ab25aded783e4e59/alacritty/src/event.rs#L573 , with getters it'll be way harder to write.

Which feature are you referring to specifically? I don't see anything there which would obviously become much more complex with the proposed change.

Reply:
> these events don't and shouldn't have public constructors.

Testing event handling will become impossible if you can't create the event structures.

Reply:
>Testing event handling will become impossible if you can't create the event structures.

That's a very good point, alacritty's event testing will be entirely broken https://github.com/alacritty/alacritty/blob/922f48e52c5643e7cce0aa72b87170d8f12b5cfa/alacritty/src/input.rs#L886 without public constructors.

Reply:
Good catch. Forwards-compatible construction could perhaps be done with setters and a `Default` impl that comes up with dummy values, or similar. Doesn't seem too terrible.

We could also just scrap forwards-compatibility concerns. Winit's not typically an interface library, especially now that we have raw-window-handle, so occasional breakage could be fine so long as nobody gets too emotionally attached to the eventual 1.0.

Reply:
Of course, there's also the option for alacritty to refactor its interfaces to use a locally-defined type that implements `From<WindowEvent>` or similar. This might be desirable from an aesthetic point of view as that it lets them get rid of the hacks used to fill in data they don't care about anyway, such as the `transmute_copy` invocations for IDs.

Reply:
Yeah, and then every application create one more event to test things, which just copies winit event, except some fields they don't care about, and then you're forced to create generic abstractions over event just to support testing here or convert everything into your event adding more clones/copies.

Like alacritty is just an example here of an app, which is using winit events for testing, to just prove that it's not something uncommon.

Reply:
Every winit app I've ever written has immediately transformed event data into some other form, typically simple primitive values passed to a handler method. There are a variety of different approaches to consuming these events; it is absolutely not the case that literally every application would be writing a bunch of redundant code.

If you have hard data on how what proportion of winit applications do, in fact, store/pass winit events around to a significant degree, that would certainly be useful for informing the design!

Reply:
By 'every application' I mean that you don't have choice anymore to do what you're doing or to do what we're doing. You just have one path - have your own event type and translate winit events.

Like ok, I personally can live with such change and we likely can refactor alacritty event handling adding even more quirks to some winit cases (though we'll likely remove one of them soon). But I'm not really sure for how long I want to workaround winit design, since it's becoming stricter and stricter.

Reply:
Those are reasonable concerns, and I've added constructors and setters for each of the event types to address those.

I'm hesitant to expose the struct fields directly, since that means that computed fields would be exposed with a different syntax than stored fields, and we're using computed fields to add an `is_up()` method that can be used instead of `!is_down()` for detecting keyup events.

Reply:
>I'm hesitant to expose the struct fields directly, since that means that computed fields would be exposed with a different syntax than stored fields, and we're using computed fields to add an is_up() method that can be used instead of !is_down() for detecting keyup events.

thx.